### PR TITLE
Run file I/O in background for Crashes service

### PR DESCRIPTION
**Pull request by jaelim-ms**

### CODE REVIEW

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 1]
```diff
@@ -142,6 +166,10 @@ private Crashes() {
         mCrashesListener = DEFAULT_ERROR_REPORTING_LISTENER;
         mUnprocessedErrorReports = new LinkedHashMap<>();
         mErrorReportCache = new LinkedHashMap<>();
+        HandlerThread thread = new HandlerThread(THREAD_NAME);
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Why using handler thread for one shot tasks?
This case is legit for ASyncTask as after we load what we need, there won't be any other use of the thread until application dies.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 2]
```diff
@@ -88,6 +97,18 @@
...
+     * Handler for background thread loop.
+     */
+    @VisibleForTesting
+    final Handler mHandler;
+
+    /**
+     * Handler for main thread loop.
+     */
+    @VisibleForTesting
+    final Handler mMainHandler;
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Code analysis warning here.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 3]
```diff
@@ -526,43 +585,50 @@ public void setWrapperSdkListener(WrapperSdkListener wrapperSdkListener) {
...
     @VisibleForTesting
-    private synchronized void handleUserConfirmation(@UserConfirmationDef int userConfirmation) {
+    private synchronized void handleUserConfirmation(@UserConfirmationDef final int userConfirmation) {
         if (mChannel == null) {
             MobileCenterLog.error(LOG_TAG, "Crashes service not initialized, discarding calls.");
             return;
         }
 
-        if (userConfirmation == DONT_SEND) {
+        mHandler.post(new Runnable() {
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Check if already in UI thread.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 2 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: It was exposed for unit test and now don't need it. Also removing the annotation.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 1 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: We are free from thread pool limitation with Handler. It is an SDK and trying not to consume resources as many as possible. Going with AsyncTask is not a problem since it is one shot tasks as you mentioned.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 3 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Not a big deal but it is just using same handler if it is already on background thread.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 4]
```diff
@@ -287,24 +287,19 @@ private synchronized boolean hasInstanceCrashedInLastSession() {
...
-                }
-                mLastCrashErrorReportListeners.add(listener);
-                break;
-            case PROCESSED:
-                listener.onSuccess(mLastSessionErrorReport);
-                break;
-            case PROCESSING_FAILED:
-                listener.onFailure();
-                break;
-            case NOT_FOUND:
```
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: That didn't help for 100% coverage. You can see some explanations in stackoverflow. Looks byte code has more cases.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 5]
```diff
@@ -102,6 +103,11 @@
     private static final CrashesListener DEFAULT_ERROR_REPORTING_LISTENER = new DefaultCrashesListener();
 
     /**
+     * Timeout for getting crash for the last session.
+     */
+    private static final int GET_LAST_SESSION_CRASH_REPORT_TIMEOUT = 10000;
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: We should wait forever for last session error report.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 6]
```diff
@@ -267,12 +283,23 @@ public static boolean hasCrashedInLastSession() {
 
     /**
      * Provides information about any available crash report from the last session, if it crashed.
+     * This method is a synchronous call and blocks UI thread.
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Javadoc: not UI thread but calling thread. If you call it in background it won't block UI thread.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 7]
```diff
@@ -267,12 +283,23 @@ public static boolean hasCrashedInLastSession() {
...
+     * Use {@link #getLastSessionCrashReportAsync(LastCrashErrorReportListener)} for asynchronous call.
+     *
+     * @return The crash report from the last session if one was set.
+     */
+    @Nullable
+    public static ErrorReport getLastSessionCrashReport() {
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Please add @WorkerThread

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 8]
```diff
@@ -182,6 +197,7 @@ private Crashes() {
         thread.start();
         mHandler = new Handler(thread.getLooper());
         mLastSessionCrashProcessingStatus = PREPARE_PROCESSING;
+        mSemaphore = new Semaphore(0);
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: This is wrong to use a shared semaphore if you have more than 1 caller. This is not the right synchronization mechanism to use in that case. Use a Condition or a CountDownLatch (I prefer CountDownLatch as it does not have finally block, finally is troublesome with code coverage).

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 9]
```diff
@@ -155,7 +157,9 @@
     /**
      * Flag indicates whether Crashes found ErrorReport for the last session.
      */
-    private boolean mFoundLastSessionErrorReport;
+    private LastCrashProcessingStatus mFoundLastSessionErrorReport;
+
+    private List<LastCrashErrorReportListener> mLastCrashErrorReportListeners;
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Set

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 10]
```diff
@@ -253,28 +258,44 @@ public static boolean hasCrashedInLastSession() {
...
+                if (mLastCrashErrorReportListeners == null) {
+                    mLastCrashErrorReportListeners = new ArrayList<>();
+                }
+                mLastCrashErrorReportListeners.add(listener);
+                break;
+            case PROCESSED:
+                listener.onSuccess(mLastSessionErrorReport);
+                break;
+            case PROCESSING_FAILED:
+                listener.onFailure();
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: We should not have a failure case for getSessionErrorReport, either return result or return null like before. The sync version has to return either a result or null and never throw an exception, so why would the async version introduce that?

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 11]
```diff
@@ -694,6 +745,37 @@ void saveErrorLog(ManagedErrorLog errorLog, File errorStorageDirectory, String f
...
+     */
+    private enum LastCrashProcessingStatus {
+        PREPARE_PROCESSING,
+        PROCESSING,
+        PROCESSED,
+        PROCESSING_FAILED,
+        NOT_FOUND
+    }
+
+    public interface LastCrashErrorReportListener {
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: This is related to another comment on simplifying cases but we should not have this interface. We will have more data to return asynchronously like that in the near future such as installId for example.

We should add a generics callback in the core module with only 1 method onResult(T data) and data can be null (if not found), and should never fail on an exception.

Azme has that generic callback if you want to look at it.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 12]
```diff
@@ -238,12 +294,61 @@ public static ErrorReport getLastSessionCrashReport() {
...
+                mLastSessionCrashProcessingStatus == PROCESSING) {
+            if (mLastCrashErrorReportListeners == null) {
+                mLastCrashErrorReportListeners = new ArrayList<>();
+            }
+            mLastCrashErrorReportListeners.add(listener);
+            MobileCenterLog.info(LOG_TAG, "Crashes for the last session have not been processed yet. The SDK will call listener when it completes process.");
+        } else if (mLastSessionCrashProcessingStatus == PROCESSED)
+            listener.onSuccess(mLastSessionErrorReport);
+        else if (mLastSessionCrashProcessingStatus == PROCESSING_FAILED)
+            listener.onFailure();
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: We should not have a failure case for getSessionErrorReport, either return result or return null like before. The sync version has to return either a result or null and never throw an exception, so why would the async version introduce that?

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 13]
```diff
@@ -613,6 +800,26 @@ void saveErrorLog(ManagedErrorLog errorLog, File errorStorageDirectory, String f
         MobileCenterLog.debug(Crashes.LOG_TAG, "Saved JSON content for ingestion into " + errorLogFile);
     }
 
+    public interface LastCrashErrorReportListener {
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: This is related to another comment on simplifying cases but we should not have this interface. We will have more data to return asynchronously like that in the near future such as installId for example.

We should add a generics callback in the core module with only 1 method onResult(T data) and data can be null (if not found), and should never fail on an exception.

Azme has that generic callback if you want to look at it.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 1 above)
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Make sure last session error report is loaded first, then process pending crashes.
Then quit looper?

**Comment by guperrot** [comment]
one of the review section is marked as outdated because I did the review from an individual commit, will redo the comments.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 12 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Introduced generic callback in mobile-center and replaced by that.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 13 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Introduced generic callback in mobile-center and replaced by that.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 7 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Added.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 6 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Agreed.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 8 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Yes, you are right. It will block forever if there are more than 1 callers. Changed the implementation by using CountDownLatch.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 5 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: It sounds like a PM question for requirement.
-> Got the same opinion. Removing timeout and just wait forever.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 1 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: We need to talk about this. Taking offline.

Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Talked offline. Will be done in a separate PR, switching to ASyncTask might be more practical to handle cancelling on disabled and then this comment would be obsolete.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 14]
```diff
@@ -219,16 +251,18 @@ public static void notifyUserConfirmation(@UserConfirmationDef int userConfirmat
...
      * @return {@code true} if a crash was recorded in the last session, otherwise {@code false}.
      */
     public static boolean hasCrashedInLastSession() {
-        return getLastSessionCrashReport() != null;
+        return getInstance().hasInstanceCrashedInLastSession();
     }
 
     /**
      * Provides information about any available crash report from the last session, if it crashed.
+     * This method is a synchronous call and blocks caller thread.
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Please add @WorkerThread in addition to that explanation.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 15]
```diff
@@ -219,16 +251,18 @@ public static void notifyUserConfirmation(@UserConfirmationDef int userConfirmat
...
+     * Use {@link #getLastSessionCrashReportAsync(ResultCallback)} for asynchronous call.
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Can use @See also.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 16]
```diff
@@ -238,12 +272,51 @@ public static ErrorReport getLastSessionCrashReport() {
...
+    /**
      * Implements {@link #getLastSessionCrashReport()} at instance level.
      */
     private synchronized ErrorReport getInstanceLastSessionCrashReport() {
+        if (mCountDownLatch != null) {
+            MobileCenterLog.debug(LOG_TAG, "Waiting for Crashes service to complete crash report for the last session.");
+            try {
+                mCountDownLatch.await();
+            } catch (InterruptedException e) {
+                MobileCenterLog.debug(LOG_TAG, "Could not get crash report for the last session.");
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: `, e);` to log the reason why.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 17]
```diff
@@ -238,12 +272,51 @@ public static ErrorReport getLastSessionCrashReport() {
...
     }
 
+    /**
+     * Implements {@link #getLastSessionCrashReportAsync(ResultCallback)} at instance level.
+     */
+    private synchronized void getInstanceLastSessionCrashReportAsync(ResultCallback<ErrorReport> callback) {
+        if (mCountDownLatch == null || mCountDownLatch.getCount() <= 0)
+            callback.onResult(mLastSessionErrorReport);
+        else {
+            if (mLastCrashErrorReportCallbacks == null) {
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Should simplify, in other places we don't lazy initialize listeners.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 18]
```diff
@@ -238,12 +272,51 @@ public static ErrorReport getLastSessionCrashReport() {
...
+                mLastCrashErrorReportCallbacks = new ArrayList<>();
+            }
+            mLastCrashErrorReportCallbacks.add(callback);
+            MobileCenterLog.info(LOG_TAG, "Crashes for the last session have not been processed yet. The SDK will call listener when it completes process.");
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: processing. (or finishes to process or equivalent)

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java [hunk 19]
```diff
@@ -408,53 +480,105 @@ private void initialize() {
...
+                    @Override
+                    public void run() {
+                        String logFileContents = StorageHelper.InternalStorage.read(logFile);
+                        if (logFileContents == null)
+                            MobileCenterLog.error(LOG_TAG, "Error reading last session error log.");
+                        else {
+                            try {
+                                ManagedErrorLog log = (ManagedErrorLog) mLogSerializer.deserializeLog(logFileContents);
+                                mLastSessionErrorReport = buildErrorReport(log);
+                                MobileCenterLog.debug(LOG_TAG, "Completed to process crash report for the last session.");
```
Comment (guperrot) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: "Completed to process" => I would suggest just "Processed"

**Comment by guperrot** [comment]
Also would suggest using both versions of the get in Sasquatch, for example, use async, and inside the async callback call the sync getter to log so that both functions are used there.

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 14 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Done

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 19 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Done

File: sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java (hunk 18 above)
Comment (jaelim-ms) [inline sdk/mobile-center-crashes/src/main/java/com/microsoft/azure/mobile/crashes/Crashes.java]: Done

**Comment by jaelim-ms** [comment]
Your last feedback will be done in the next PR.

**Comment by jaelim-ms** [comment]
I meant the example will be in the next PR. As we introduce a new API, we should show how to use in our test application. This will also fix lint issue. Anyway, it will come soon. 😄

**Comment by guperrot** [comment]
Ah ok, misread the comment "next pr" ;)
